pdns/dnsdistdist/dnsdist-secpoll.cc

```diff
@@ -0,0 +1,249 @@
+        warnlog("Error while waiting for the secpoll response from stub resolver %s: %d", dest.toString(), ret);
+      }
+      continue;
+    }
+    else if (ret == 0) {
+      if (g_verbose) {
+        warnlog("Timeout while waiting for the secpoll response from stub resolver %s", dest.toString());
+      }
+      continue;
+    }
+
+    try {
+      sock.read(reply);
+    }
+    catch(const std::exception& e) {
+      if (g_verbose) {
+        warnlog("Error while reading for the secpoll response from stub resolver %s: %s", dest.toString(), e.what());
+      }
+      continue;
+    }
+
+    if (reply.size() <= sizeof(struct dnsheader)) {
+      if (g_verbose) {
+        warnlog("Too short answer of size %d received from the secpoll stub resolver %s", reply.size(), dest.toString());
+      }
+      continue;
+    }
+
+    struct dnsheader d;
+    memcpy(&d, reply.c_str(), sizeof(d));
+    if (d.id != pw.getHeader()->id) {
+      if (g_verbose) {
+        warnlog("Invalid ID (%d / %d) received from the secpoll stub resolver %s", d.id, pw.getHeader()->id, dest.toString());
+      }
+      continue;
+    }
+
+    if (d.rcode != RCode::NoError) {
+      if (g_verbose) {
+        warnlog("Response code '%s' received from the secpoll stub resolver %s for '%s'", RCode::to_s(d.rcode), dest.toString(), queriedName);
+      }
+
+      /* no need to try another resolver if the domain does not exist */
+      if (d.rcode == RCode::NXDomain) {
+        throw std::runtime_error("Unable to get a valid Security Status update");
+      }
+      continue;
+    }
+
+    if (ntohs(d.qdcount) != 1 || ntohs(d.ancount) != 1) {
+      if (g_verbose) {
+        warnlog("Invalid answer (qdcount %d / ancount %d) received from the secpoll stub resolver %s", dest.toString());
+      }
+      continue;
+    }
+
+    uint16_t receivedType;
+    uint16_t receivedClass;
+    DNSName receivedName(reply.c_str(), reply.size(), sizeof(dnsheader), false, &receivedType, &receivedClass);
+
+    if (receivedName != sentName || receivedType != QType::TXT || receivedClass != QClass::IN) {
+      if (g_verbose) {
+        warnlog("Invalid answer, either the qname (%s / %s), qtype (%s / %s) or qclass (%d / %d) does not match, received from the secpoll stub resolver %s", receivedName, sentName, QType(receivedType).getName(), QType(QType::TXT).getName(), receivedClass, QClass::IN, dest.toString());
+      }
+      continue;
+    }
+
+    return getFirstTXTAnswer(reply);
+  }
+
+  throw std::runtime_error("Unable to get a valid Security Status update");
+}
+
+static bool g_secPollDone{false};
+std::string g_secPollSuffix{"secpoll.powerdns.com."};
+time_t g_secPollInterval{3600};
+
+void doSecPoll(const std::string& suffix)
+{
+  if (suffix.empty()) {
+    return;
+  }
+
+  const std::string pkgv(PACKAGEVERSION);
+  bool releaseVersion = pkgv.find("0.0.") != 0;
+
+  struct timeval now;
+  gettimeofday(&now, 0);
+
+  const std::string version = "dnsdist-" + std::string(PACKAGEVERSION);
```

Comment:
`s/std::string(PACKAGEVERSION)/pkgv/` ?

Reply:
Done! Also removed a useless call to `gettimeofday()`.